mir_eval/display.py

```diff
@@ -0,0 +1,819 @@
+
+    height : scalar or np.ndarray, shape=(n,), optional
+        Height for each label.
+        If scalar, the same value is applied to all labels.
+        By default, each label has `height=1`.
+
+    extend_labels : bool
+        If `False`, only values of `labels` that also exist in `label_set`
+        will be shown.
+
+        If `True`, all labels are shown, with those in `labels` but
+        not in `label_set` appended to the top of the plot.
+        A horizontal line is drawn to indicate the separation between
+        values in or out of `label_set`.
+
+    ax : matplotlib.pyplot.axes
+        An axis handle on which to draw the intervals.
+        If none is provided, a new set of axes is created.
+
+    tick : bool
+        If `True`, sets tick positions and labels on the y-axis.
+
+    kwargs
+        Additional keyword arguments to pass to
+        `matplotlib.patches.Rectangle`.
+
+    Returns
+    -------
+    ax : matplotlib.pyplot.axes._subplots.AxesSubplot
+        A handle to the (possibly constructed) plot axes
+    '''
+
+    # Get the axes handle
+    ax, _ = __get_axes(ax=ax)
+
+    if label_set is None:
+        # If we have non-empty pre-existing tick labels, use them
+        label_set = [_.get_text() for _ in ax.get_yticklabels()]
+        # If none of the label strings have content, treat it as empty
+        if not any(label_set):
+            label_set = []
+    else:
+        label_set = list(label_set)
+
+    # Put additional labels at the end, in order
+    if extend_labels:
+        ticks = label_set + sorted(set(labels) - set(label_set))
+    elif label_set:
+        ticks = label_set
+    else:
+        ticks = sorted(set(labels))
+
+    seg_def_style = dict(linewidth=1)
+
+    seg_map = dict()
+    seg_y = dict()
+
+    style = next(ax._get_patches_for_fill.prop_cycler)
+    if base is None:
+        base = np.arange(len(ticks))
+
+    if height is None:
+        height = 1
+
+    if np.isscalar(height):
+        height = height * np.ones_like(base)
+
+    for y0, yi, lab in zip(base, height, ticks):
+        seg_map[lab] = dict(label=lab)
+        seg_map[lab].update(seg_def_style)
+        seg_map[lab].update(style)
+        # Swap color->facecolor to preserve edge color on rects
+        seg_map[lab]['facecolor'] = seg_map[lab].pop('color')
+        seg_map[lab].update(kwargs)
+
+        seg_y[lab] = (y0, yi)
+
+    seen = set()
+    for ival, lab in zip(intervals, labels):
+        if lab not in seg_map:
+            continue
+        over_lab = seg_map[lab].get('label', lab)
+
+        # If we've already seen this label, remove it
+        # This way, it only appears once in the legend
+        if over_lab in seen:
+            seg_map[lab].pop('label', None)
+
+        seen.add(over_lab)
+        ax.add_patch(Rectangle((ival[0], seg_y[lab][0]),
+                               ival[1] - ival[0],
+                               seg_y[lab][1], **seg_map[lab]))
```

Comment:
I rewrote `labeled_intervals` using `BrokenBarHCollection`, since `PatchCollection` does not propagate labels correctly.

It's considerably simpler and faster now.